manta-crypto/src/eclair/cmp.rs

```diff
@@ -35,9 +40,23 @@ where
     {
         self.eq(other, compiler).not(compiler)
     }
+
+    /// Asserts that `self` and `rhs` are equal.
+    ///
+    /// # Implementation Note
+    ///
+    /// This method is an optimization path for the case when comparing for equality and then
+    /// asserting is more expensive than a custom assertion.
+    #[inline]
+    fn assert_equal(&self, rhs: &Rhs, compiler: &mut COM)
+    where
+        COM: Assert,
+    {
+        let are_equal = self.eq(rhs, compiler);
+        compiler.assert(&are_equal);
+    }
```

Comment:
I don't see how calling the new assert_equal for AssertEq which calls an assert_equal defined for PartialEq, which in turn is the old assert_equal for AssertEq can be more efficient than directly calling the old one. Why is that?

Reply:
Here we are allowing the specific type to customize the behavior of `assert_eq` whenever it can be more efficient than the default behavior. If we left this implementation to the compiler, it would have to provide an explicit implementation that worked over all types and would have to be generic over all of them. Instead, we let the type handle this behavior when it knows how to do this better. Here's an example:
```rust
struct Struct<A, B> {
    a: A,
    b: B,
}

impl<A, B, COM> PartialEq<Self, COM> for Struct<A, B>
where
    A: PartialEq<Self, COM>,
    B: PartialEq<Self, COM>,
{
    fn assert_equal(&self, rhs: &Self, compiler: &mut COM) {
        compiler.assert_eq(self.a, rhs.a);
        compiler.assert_eq(self.b, rhs.b);
    }
}
```
instead of
```rust
fn assert_equal(&self, rhs: &Self, compiler: &mut COM) {
    let are_equal = self.a.eq(rhs.a, compiler).bitand(self.b.eq(rhs.b, compiler), compiler);
    compiler.assert(are_equal);
}
```
Notice that in this second one we have to construct the intermediate boolean `are_equal` and we have to pay for a `bitand` operation as well.